The `Alignment` class will compute the optimal global pairwise alignment of the two genomic sequences found in the two FASTA files. <br/>

Functions filled out for the `Alightnment` class: 
* `localAlignment()`  
* `setMatrixScoring(String fileName)`
* `getScore(char c1, char c2)`
* `getGapScore(int i, int j, boolean isGapAbove)`
* `globalAlignment_FAST()`
* `getVectorMean(Vector<Integer> v)`
* `getVectorStandardDeviation(Vector<Integer> v)`
* `calculatePValue()`
